DataFormats/Provenance/src/IndexIntoFile.cc

```diff
@@ -1980,27 +2030,408 @@ namespace edm {
+      // If first entry for a lumi
+      if (!runOrLumiEntry.isRun() && runOrLumiEntry.orderPHIDRunLumi() != previousLumi) {
+        previousLumi = runOrLumiEntry.orderPHIDRunLumi();
+
+        // Fill map holding the first index into runOrLumiEntries for each lum
+        info.firstIndexOfLumi_[runOrLumiEntry.orderPHIDRunLumi()] = index;
+      }
+
+      // If first entry for a run
+      if (runOrLumiEntry.orderPHIDRun() != previousRun) {
+        previousRun = runOrLumiEntry.orderPHIDRun();
+
+        // Fill map holding the first index into runOrLumiEntries for each run
+        info.firstIndexOfRun_[runOrLumiEntry.orderPHIDRun()] = index;
+
+        // Look ahead to see if the run has events or not
+        bool runHasEvents = false;
+        for (int indexWithinRun = index + 1;
+             indexWithinRun < iEnd && runOrLumiEntries[indexWithinRun].orderPHIDRun() == runOrLumiEntry.orderPHIDRun();
+             ++indexWithinRun) {
+          if (runOrLumiEntries[indexWithinRun].beginEvents() != invalidEntry) {
+            runHasEvents = true;
+            break;
+          }
+        }
+        if (!runHasEvents) {
+          info.runsWithNoEvents_.push_back({runOrLumiEntry.entry(), index});
+        }
+      }
+      ++index;
+    }
+
+    std::sort(info.runsWithNoEvents_.begin(),
+              info.runsWithNoEvents_.end(),
+              [](TTreeEntryAndIndex const& left, TTreeEntryAndIndex const& right) -> bool {
+                return left.ttreeEntry_ < right.ttreeEntry_;
+              });
+
+    info.nextRunWithNoEvents_ = info.runsWithNoEvents_.cbegin();
+    info.endRunsWithNoEvents_ = info.runsWithNoEvents_.cend();
+  }
+
+  void IndexIntoFile::IndexIntoFileItrEntryOrder::fillIndexesSortedByEventEntry(EntryOrderInitializationInfo& info) {
+    auto const& runOrLumiEntries = indexIntoFile()->runOrLumiEntries();
+
+    int index = 0;
+    for (auto const& runOrLumiEntry : runOrLumiEntries) {
+      if (runOrLumiEntry.beginEvents() != invalidEntry) {
+        info.indexesSortedByEventEntry_.push_back({runOrLumiEntry.beginEvents(), index});
+      }
+      ++index;
+    }
+
+    std::sort(info.indexesSortedByEventEntry_.begin(),
+              info.indexesSortedByEventEntry_.end(),
+              [](TTreeEntryAndIndex const& left, TTreeEntryAndIndex const& right) -> bool {
+                return left.ttreeEntry_ < right.ttreeEntry_;
+              });
+
+    // The next "for loop" is just a sanity check, it should always pass.
+    int previousIndex = invalidIndex;
+    for (auto const& eventSequence : info.indexesSortedByEventEntry_) {
+      int currentIndex = eventSequence.runOrLumiIndex_;
+      if (previousIndex != invalidIndex) {
+        assert(runOrLumiEntries[previousIndex].endEvents() == runOrLumiEntries[currentIndex].beginEvents());
+      }
+      previousIndex = currentIndex;
+    }
+  }
+
+  void IndexIntoFile::IndexIntoFileItrEntryOrder::fillIndexesToLastContiguousEvents(EntryOrderInitializationInfo& info) {
```

Comment:
Same here

Reply:
Done. I moved it into the Info class. That is better. Thanks.